File: testdata/jennies/builders/foreign_builder/PythonBuilder/builders/builder_pkg.py
import typing
from ..cog import builder as cogbuilder
from ..models import some_pkg


class SomeNiceBuilder(cogbuilder.Builder[some_pkg.SomeStruct]):
    _internal: some_pkg.SomeStruct

    def __init__(self):
        self._internal = some_pkg.SomeStruct()

    def build(self) -> some_pkg.SomeStruct:
        """
        Builds the object.
        """
        return self._internal    
    
    def title(self, title: str) -> typing.Self:    
        self._internal.title = title
    
        return self
    

File: testdata/jennies/builders/foreign_builder/RustBuilder/src/builders/builder_pkg.rs
use crate::cog;
use crate::types::some_pkg;

#[derive(Debug, Clone)]
pub struct SomeNiceBuilderBuilder {
    internal: some_pkg::SomeStruct,
    pub(crate) errors: Vec<cog::BuildError>,
}

impl SomeNiceBuilderBuilder {
    pub fn new() -> Self {
        Self {
            internal: some_pkg::SomeStruct::default(),
            errors: Vec::new(),
        }
    }
}

impl Default for SomeNiceBuilderBuilder {
    fn default() -> Self {
        Self::new()
    }
}

impl SomeNiceBuilderBuilder {
    pub fn title(mut self, title: String) -> Self {
        self.internal.title = title;

        self
    }
}

impl cog::Builder<some_pkg::SomeStruct> for SomeNiceBuilderBuilder {
    fn build(&self) -> Result<some_pkg::SomeStruct, Vec<cog::BuildError>> {
        if !self.errors.is_empty() {
            return Err(self.errors.clone());
        }

        Ok(self.internal.clone())
    }
}
